--- build/boogie.ebnf.orig	2026-03-21 12:53:38.387720994 +1000
+++ build/boogie.ebnf	2026-03-21 12:54:43.834931155 +1000
@@ -11,7 +11,7 @@
   ;
   ; productions
   ;
-  BoogiePL ::=  ( Consts | Function | Axiom | UserDefinedTypes | Datatype | GlobalVars | InvariantDecl | "yield" ( InvariantDecl | YieldProcedureDecl ) | Pure ( Procedure | ActionDecl ) | Implementation )*
+  BoogiePL ::=  ( Consts | Function | Axiom | UserDefinedTypes | Datatype | GlobalVars | InvariantDecl | "yield" ( InvariantDecl | YieldProcedureDecl ) | ( Pure )? ( Procedure | ActionDecl ) | Implementation )*
   Consts ::=  "const" Attribute* ( "unique" )? IdsType ( ";" | "uses" "{" ( Axiom )* "}" )
   Function ::=  ( "revealed" )? "function" Attribute* Ident TypeParams? "(" ( VarOrType ( "," VarOrType )* )? ")" ( "returns" "(" VarOrType ")" | ":" Type ) ( "{" Expression "}" ( "uses" "{" ( Axiom )* "}" )? | "uses" "{" ( Axiom )* "}" | ";" )
   Axiom ::=  ( "hideable" )? "axiom" Attribute* Proposition ";"
@@ -20,7 +20,7 @@
   GlobalVars ::=  "var" Attribute* IdsTypeWheres ";"
   InvariantDecl ::=  "invariant" Attribute* Ident ProcFormals ";" Invariant*
   YieldProcedureDecl ::=  MoverQualifier? "procedure" Attribute* Ident ProcFormals ( "returns" ProcFormals )? ( ";" SpecYieldPrePost* | SpecYieldPrePost* ImplBody )
-  Pure ::=  ( "pure" )?
+  Pure ::=  ( "pure" )
   Procedure ::=  "procedure" ProcSignature ( ";" Spec* | Spec* ImplBody )
   ActionDecl ::=  MoverQualifier? "action" Attribute* Ident ProcFormals ( "returns" ProcFormals )? ( ";" SpecAction* | SpecAction* ImplBody )
   Implementation ::=  "implementation" ProcSignature ImplBody
@@ -81,7 +81,7 @@
   _ImpliesExpression ::=  _LogicalExpression  | ImpliesExpression
   ImpliesExpression ::=   _LogicalExpression  (  ImpliesOp _ImpliesExpression | ExpliesOp _LogicalExpression ( ExpliesOp _LogicalExpression )* )
   EquivOp ::=  "<==>" | "\u21d4"
-  _LogicalExpression ::=  _RelationalExpression  | LogicalExpression
+  _LogicalExpression ::=  _RelationalExpression  | LogicalExpression | IfThenElseExpression
   LogicalExpression ::=   _RelationalExpression  (  AndOp _RelationalExpression ( AndOp _RelationalExpression )* | OrOp _RelationalExpression ( OrOp _RelationalExpression )* )
   ImpliesOp ::=  "==>" | "\u21d2"
   ExpliesOp ::=  "<==" | "\u21d0"
@@ -110,7 +110,7 @@
   _ArrayExpression ::=  AtomExpression  | ArrayExpression
   ArrayExpression ::=   AtomExpression  (  "[" ( Expression ( "," Expression )* ( ":=" Expression )? | ":=" Expression )? "]" | "->" ( Ident | "(" Ident ":=" Expression ")" ) )+
   Nat ::=  digits
-  AtomExpression ::=  ( "false" | "true" | ( "roundNearestTiesToEven" | "RNE" ) | ( "roundNearestTiesToAway" | "RNA" ) | ( "roundTowardPositive" | "RTP" ) | ( "roundTowardNegative" | "RTN" ) | ( "roundTowardZero" | "RTZ" ) | Nat | Dec | Float | BvLit | string | Ident ( "(" ( Expressions )? ")" )? | "old" "(" Expression ")" | "int" "(" Expression ")" | "real" "(" Expression ")" | "(" ( Expression | Forall QuantifierBody | Exists QuantifierBody | Lambda QuantifierBody | LetExpr ) ")" | IfThenElseExpression | CodeExpression )
+  AtomExpression ::=  ( "false" | "true" | ( "roundNearestTiesToEven" | "RNE" ) | ( "roundNearestTiesToAway" | "RNA" ) | ( "roundTowardPositive" | "RTP" ) | ( "roundTowardNegative" | "RTN" ) | ( "roundTowardZero" | "RTZ" ) | Nat | Dec | Float | BvLit | string | Ident ( "(" ( Expressions )? ")" )? | "old" "(" Expression ")" | "int" "(" Expression ")" | "real" "(" Expression ")" | "(" ( Expression | Forall QuantifierBody | Exists QuantifierBody | Lambda QuantifierBody | LetExpr ) ")" | CodeExpression )
   Dec ::=  ( decimal | dec_float )
   Float ::=  float
   BvLit ::=  bvlit
@@ -119,11 +119,11 @@
   Exists ::=  "exists" | "\u2203"
   Lambda ::=  "lambda" | "\u03bb"
   LetExpr ::=  "var" LetVar ( "," LetVar )* ":=" Expression ( "," Expression )* ";" Attribute* Expression
-  IfThenElseExpression ::=  "if" Expression "then" Expression "else" Expression
+  IfThenElseExpression ::=  "if" Expression "then" Expression "else" _LogicalExpression
   CodeExpression ::=  "|{" LocalVars* SpecBlock ( SpecBlock )* "}|"
   SpecBlock ::=  Ident ":" ( LabelOrCmd )* ( "goto" Attribute* Idents | "return" Attribute* Expression ) ";"
   AttributeOrTrigger ::=  "{" ( ":" Ident ( AttributeParameter ( "," AttributeParameter )* )? | Expression ( "," Expression )* ) "}"
-  AttributeParameter ::=  ( string | Expression )
+  AttributeParameter ::=  ( Expression )
   QSep ::=  "::" | "\u2022"
   LetVar ::=  Attribute* Ident
   ;
