--- build/boogie.ebnf.orig	2026-03-21 14:11:55.940029190 +1000
+++ build/boogie.ebnf	2026-03-21 14:13:38.097658930 +1000
@@ -11,7 +11,7 @@
   ;
   ; productions
   ;
-  BoogiePL ::=  ( Consts | Function | Axiom | UserDefinedTypes | Datatype | GlobalVars | InvariantDecl | "yield" ( InvariantDecl | YieldProcedureDecl ) | Pure ( Procedure | ActionDecl ) | Implementation )*
+  BoogiePL ::=  ( Consts | Function | Axiom | UserDefinedTypes | Datatype | GlobalVars | InvariantDecl | "yield" ( InvariantDecl | YieldProcedureDecl ) | ( Pure )? ( Procedure | ActionDecl ) | Implementation )*
   Consts ::=  "const" Attribute* ( "unique" )? IdsType ( ";" | "uses" "{" ( Axiom )* "}" )
   Function ::=  ( "revealed" )? "function" Attribute* Ident TypeParams? "(" ( VarOrType ( "," VarOrType )* )? ")" ( "returns" "(" VarOrType ")" | ":" Type ) ( "{" _Expression "}" ( "uses" "{" ( Axiom )* "}" )? | "uses" "{" ( Axiom )* "}" | ";" )
   Axiom ::=  ( "hideable" )? "axiom" Attribute* Proposition ";"
@@ -20,7 +20,7 @@
   GlobalVars ::=  "var" Attribute* IdsTypeWheres ";"
   InvariantDecl ::=  "invariant" Attribute* Ident ProcFormals ";" Invariant*
   YieldProcedureDecl ::=  MoverQualifier? "procedure" Attribute* Ident ProcFormals ( "returns" ProcFormals )? ( ";" SpecYieldPrePost* | SpecYieldPrePost* ImplBody )
-  Pure ::=  ( "pure" )?
+  Pure ::=  ( "pure" )
   Procedure ::=  "procedure" ProcSignature ( ";" Spec* | Spec* ImplBody )
   ActionDecl ::=  MoverQualifier? "action" Attribute* Ident ProcFormals ( "returns" ProcFormals )? ( ";" SpecAction* | SpecAction* ImplBody )
   Implementation ::=  "implementation" ProcSignature ImplBody
@@ -36,8 +36,8 @@
   AttributesIdsTypeWhere ::=  Attribute* _IdsTypeWhere
   _IdsTypeWhere ::=  Idents ":" Type  | IdsTypeWhere
   IdsTypeWhere ::=   Idents ":" Type  (  "where" _Expression )
-  _Expression ::=  _ImpliesExpression  | Expression
-  Expression ::=   _ImpliesExpression  (  EquivOp _ImpliesExpression )+
+  _Expression ::=  IfThenElseExpression | _ImpliesExpression  | EquivExpression
+  EquivExpression ::=   _ImpliesExpression  (  EquivOp _ImpliesExpression )+
   TypeAtom ::=  ( "int" | "real" | "bool" | "(" Type ")" )
   Ident ::=  ( ident | "atomic" | "both" | "left" | "right" | "reveal" | "hide" | "push" | "pop" )
   TypeArgs ::=  ( TypeAtom TypeArgs? | Ident TypeArgs? | MapType )
@@ -112,7 +112,7 @@
   _ArrayExpression ::=  AtomExpression  | ArrayExpression
   ArrayExpression ::=   AtomExpression  (  "[" ( _Expression ( "," _Expression )* ( ":=" _Expression )? | ":=" _Expression )? "]" | "->" ( Ident | "(" Ident ":=" _Expression ")" ) )+
   Nat ::=  digits
-  AtomExpression ::=  ( "false" | "true" | ( "roundNearestTiesToEven" | "RNE" ) | ( "roundNearestTiesToAway" | "RNA" ) | ( "roundTowardPositive" | "RTP" ) | ( "roundTowardNegative" | "RTN" ) | ( "roundTowardZero" | "RTZ" ) | Nat | Dec | Float | BvLit | string | Ident ( "(" ( Expressions )? ")" )? | "old" "(" _Expression ")" | "int" "(" _Expression ")" | "real" "(" _Expression ")" | "(" ( _Expression | Forall QuantifierBody | Exists QuantifierBody | Lambda QuantifierBody | LetExpr ) ")" | IfThenElseExpression | CodeExpression )
+  AtomExpression ::=  ( "false" | "true" | ( "roundNearestTiesToEven" | "RNE" ) | ( "roundNearestTiesToAway" | "RNA" ) | ( "roundTowardPositive" | "RTP" ) | ( "roundTowardNegative" | "RTN" ) | ( "roundTowardZero" | "RTZ" ) | Nat | Dec | Float | BvLit | string | Ident ( "(" ( Expressions )? ")" )? | "old" "(" _Expression ")" | "int" "(" _Expression ")" | "real" "(" _Expression ")" | "(" ( _Expression | Forall QuantifierBody | Exists QuantifierBody | Lambda QuantifierBody | LetExpr ) ")" | CodeExpression )
   Dec ::=  ( decimal | dec_float )
   Float ::=  float
   BvLit ::=  bvlit
@@ -125,7 +125,7 @@
   CodeExpression ::=  "|{" LocalVars* SpecBlock ( SpecBlock )* "}|"
   SpecBlock ::=  Ident ":" ( LabelOrCmd )* ( "goto" Attribute* Idents | "return" Attribute* _Expression ) ";"
   AttributeOrTrigger ::=  "{" ( ":" Ident ( AttributeParameter ( "," AttributeParameter )* )? | _Expression ( "," _Expression )* ) "}"
-  AttributeParameter ::=  ( string | _Expression )
+  AttributeParameter ::=  ( _Expression )
   QSep ::=  "::" | "\u2022"
   LetVar ::=  Attribute* Ident
   ;
